File: android-app/app/src/main/java/com/rakshak/ai/engine/UpiQrDetector.kt
package com.rakshak.ai.engine

import java.net.URLDecoder

data class UpiResult(
    val isThreat: Boolean,
    val riskScore: Int,
    val vpa: String,
    val payeeName: String,
    val amount: Double?,
    val transactionNote: String,
    val riskReasons: List<String>,
    val hasDeceptiveIntent: Boolean,
    val frictionMessageEn: String,
    val frictionMessageHi: String
)

object UpiQrDetector {

    private val DECEPTIVE_KEYWORDS = listOf(
        "cashback", "reward", "refund", "winner", "lottery", "prize", "gift", "claim", "bonus"
    )

    fun parseUpiPayload(payload: String, isInBloom: Boolean = false): UpiResult? {
        val raw = payload.trim()
        if (raw.isEmpty()) return null

        var vpa = ""
        var payeeName = "Unknown Merchant / Receiver"
        var amount: Double? = null
        var transactionNote = ""

        if (raw.startsWith("upi://pay")) {
            val query = raw.substringAfter("?", "")
            val params = query.split("&")
            for (param in params) {
                val parts = param.split("=", limit = 2)
                if (parts.size != 2) continue
                val key = parts[0].lowercase()
                val value = safeDecode(parts[1])
                when (key) {
                    "pa" -> vpa = value
                    "pn" -> payeeName = value
                    "am" -> amount = value.toDoubleOrNull()
                    "tn" -> transactionNote = value
                }
            }
        } else if (raw.contains("@")) {
            vpa = raw
        } else {
            return null
        }

        val hasDeceptiveIntent = DECEPTIVE_KEYWORDS.any { kw ->
            vpa.contains(kw, ignoreCase = true) ||
                payeeName.contains(kw, ignoreCase = true) ||
                transactionNote.contains(kw, ignoreCase = true)
        }

        var riskScore = 10
        val riskReasons = mutableListOf<String>()

        if (isInBloom) {
            riskScore = 99
            riskReasons.add("VPA is listed on the National Fraud & Sybil Registry.")
        }
        if (hasDeceptiveIntent) {
            riskScore = maxOf(riskScore, 92)
            riskReasons.add("Deceptive 'Cashback/Reward' name detected to lure user into completing a payment.")
        }
        if (amount != null && hasDeceptiveIntent) {
            riskScore = maxOf(riskScore, 98)
            riskReasons.add("Automated ₹${amount} DEBIT transaction disguised as reward claim.")
        }

        val isThreat = riskScore >= 70

        return UpiResult(
            isThreat = isThreat,
            riskScore = riskScore,
            vpa = vpa,
            payeeName = payeeName,
            amount = amount,
            transactionNote = transactionNote,
            riskReasons = riskReasons,
            hasDeceptiveIntent = hasDeceptiveIntent,
            frictionMessageEn = if (amount != null)
                "DANGER: This QR/link will DEDUCT ₹$amount from YOUR bank account. UPI PIN is NEVER needed to receive money!"
            else
                "DANGER: You are about to initiate a payment to $payeeName. You cannot receive money by scanning this code!",
            frictionMessageHi = if (amount != null)
                "सावधान: यह क्यूआर कोड स्कैन करने पर आपके खाते से ₹$amount कट जाएंगे। पैसे प्राप्त करने के लिए कभी भी पिन डालने की आवश्यकता नहीं होती!"
            else
                "सावधान: आप $payeeName को पैसे भेज रहे हैं। यह कोड स्कैन करके आपको कोई पैसा या कैशबैक नहीं मिलेगा!"
        )
    }

    private fun safeDecode(value: String): String {
        return try {
            URLDecoder.decode(value, "UTF-8")
        } catch (e: Exception) {
            value
        }
    }
}

File: src/cyber/engine/upiQrDetector.js
// UPI Deep-Link Schema Inspector & Micro-Friction Intent Gate
import { globalBloomFilter } from './bloomFilter.js';

export function parseUpiPayload(payload) {
  if (!payload || typeof payload !== 'string') {
    return null;
  }

  const raw = payload.trim();
  let params = {};
  let isUpiScheme = false;

  if (raw.startsWith('upi://pay')) {
    isUpiScheme = true;
    const queryString = raw.split('?')[1] || '';
    const searchParams = new URLSearchParams(queryString);

    searchParams.forEach((val, key) => {
      params[key.toLowerCase()] = val;
    });
  } else if (raw.includes('@')) {
    // Standalone VPA entered directly
    params.pa = raw;
  }

  const vpa = params.pa || '';
  const payeeName = params.pn || 'Unknown Merchant / Receiver';
  const amount = params.am ? parseFloat(params.am) : null;
  const transactionNote = params.tn || '';
  const currency = params.cu || 'INR';

  // Check Bloom Filter for malicious VPAs
  const bloomResult = globalBloomFilter.contains(vpa);

  // Check deceptive note or payee name indicating "Cashback / Refund / Prize"
  const deceptiveKeywords = ['cashback', 'reward', 'refund', 'winner', 'lottery', 'prize', 'gift', 'claim', 'bonus'];
  const hasDeceptiveIntent = deceptiveKeywords.some(
    kw => vpa.toLowerCase().includes(kw) || payeeName.toLowerCase().includes(kw) || transactionNote.toLowerCase().includes(kw)
  );

  let riskScore = 10;
  let riskReasons = [];

  if (bloomResult.match) {
    riskScore = 99;
    riskReasons.push("VPA is listed on the National Fraud & Sybil Registry.");
  }

  if (hasDeceptiveIntent) {
    riskScore = Math.max(riskScore, 92);
    riskReasons.push("Deceptive 'Cashback/Reward' name detected to lure user into completing a payment.");
  }

  // If amount is set in a payment link sent to a user claiming to give money:
  if (amount && hasDeceptiveIntent) {
    riskScore = Math.max(riskScore, 98);
    riskReasons.push(`Automated ₹${amount} DEBIT transaction disguised as reward claim.`);
  }

  const isThreat = riskScore >= 70;

  return {
    rawPayload: raw,
    isUpiScheme,
    vpa,
    payeeName,
    amount,
    currency,
    transactionNote,
    isThreat,
    riskScore,
    riskReasons,
    hasDeceptiveIntent,
    bloomLookupMs: bloomResult.lookupTimeMs,
    // Critical warning explaining that UPI PIN = Sending Money
    actionType: "DEBIT_PAYMENT",
    frictionMessage: {
      en: amount 
        ? `DANGER: This QR/link will DEDUCT ₹${amount} from YOUR bank account. UPI PIN is NEVER needed to receive money!` 
        : `DANGER: You are about to initiate a payment to ${payeeName}. You cannot receive money by scanning this code!`,
      hi: amount
        ? `सावधान: यह क्यूआर कोड स्कैन करने पर आपके खाते से ₹${amount} कट जाएंगे। पैसे प्राप्त करने के लिए कभी भी पिन डालने की आवश्यकता नहीं होती!`
        : `सावधान: आप ${payeeName} को पैसे भेज रहे हैं। यह कोड स्कैन करके आपको कोई पैसा या कैशबैक नहीं मिलेगा!`
    }
  };
}
